### ISSUE TITLE
Adding custom metrics to YOLO model in training and validation

### ISSUE THREAD

**Issue by joortif** [ultralytics/ultralytics#17879]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm trying to add a callback method in order to calculate additional metrics (Dice Score, Precision and Recall) for different YOLO image segmentation model versions (v8, v9 and v11) after each epoch. The thing is, i want to evaluate this new metrics while using train() and valid() (for fine-tuning) methods but i'm unable to find the necessary segmentation masks resulting from each epoch from the model (to compare them with the ground truth).

I have also tried to calculate these metrics once the whole training (and validation) process is finished but the result type is a SegmentMetrics in both cases. I haven't found any information about what i'm trying to do (maybe it's not even possible or it doesn't even make sense??) but I'm trying to study the efficency of these models with different datasets and comparing them with other image segmentation models (like U-net and DeepLabv3), which is why i need to calculate a wide variety of metrics common for each model.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @joortif, thank you for your interest in Ultralytics 🚀! We recommend referring to the [Docs](https://docs.ultralytics.com) for guidance, where you can find comprehensive resources that may help with your use case, including details on [training](https://docs.ultralytics.com/usage/python/#training-python) and [validation](https://docs.ultralytics.com/usage/python/#validation-python) processes for YOLO models.

If this is a 🐛 Bug Report, please provide a clear [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us investigate and debug the issue more efficiently.

If this is a custom training ❓ Question, as it appears to be, kindly include any additional details such as relevant scripts, snippets of your callback implementation, model/dataset configurations, or logs that can help us understand your challenge better. Additionally, you can review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to enhance your training workflow. 

For further assistance, you might find it helpful to engage with our community:
- Join real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧.
- Post in-depth questions or explore existing threads on [Discourse](https://community.ultralytics.com).
- Engage with peers on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Ensure you're running the latest release of `ultralytics`, which often includes new features and fixes that might resolve existing issues. Update your package in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) as follows:

```bash
pip install -U ultralytics
```

For comparing different models and metrics, running experiments in verified environments with proper dependencies is crucial:

## Verified Environments

YOLO can be run in many verified environments for experimentation:
- Notebooks with free GPU:
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> 
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> 
  <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- [Google Cloud Deep Learning VM](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- [Amazon Deep Learning AMI](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests ensure that all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) function as expected on macOS, Windows, and Ubuntu environments.

This is an automated response to help streamline your inquiry, but rest assured, an Ultralytics engineer will follow up soon to provide additional support 😊.

**Comment by Y-T-G** [comment]
You can access the ground truth and predictions in a callback:
```python
def val_cb(validator):
     frame = inspect.currentframe().f_back.f_back
     v = frame.f_locals
     v["batch"] # GT
     v["preds"] # Predictions
model.add_callback("on_val_batch_end", val_cb)
```

But I don't understand why you want to compare it with U-Net and DeepLab. They're both semantic segmentation models while YOLO-Seg is instance segmentation model.
